clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -643,10 +644,19 @@ private KafkaConsumer(ConsumerConfig config,
             List<ConsumerInterceptor<K, V>> interceptorList = (List) (new ConsumerConfig(userProvidedConfigs)).getConfiguredInstances(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,
                     ConsumerInterceptor.class);
             this.interceptors = interceptorList.isEmpty() ? null : new ConsumerInterceptors<>(interceptorList);
+
+            // the maximum time we can block in a call to NetworkClient.poll is limited by the frequency
+            // of periodic events (heartbeats and autocommits)
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            int autoCommitIntervalMs = config.getInt(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG);
+            int heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+            this.maxPollTimeoutMs = (enableAutoCommit ? Math.min(autoCommitIntervalMs, heartbeatIntervalMs) : heartbeatIntervalMs) / 2;
```

Comment:
I understand @ewencp 's motivation for not going to the limit and dividing by 2 instead, but I feel it may be confusing to people who use heartbeat interval say `800` and call `poll(500)` but realize it always returns in 400ms but with no records. I'd prefer not to divide by a magic number but allow logic simplicity for users to reason.